beacon/light/request/scheduler.go

```diff
@@ -0,0 +1,371 @@
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	server := newServer(rs, s.clock)
+	s.servers[server] = struct{}{}
+	s.eventLock.Lock()
+	s.addEvent(Event{Type: EvRegistered, Server: server})
+	s.eventLock.Unlock()
+	server.subscribe(func(event Event) {
+		event.Server = server
+		s.eventLock.Lock()
+		s.addEvent(event)
+		s.eventLock.Unlock()
+	})
+}
+
+// UnregisterServer removes a registered server.
+func (s *Scheduler) UnregisterServer(rs requestServer) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for server := range s.servers {
+		if sl, ok := server.(*serverWithLimits); ok && sl.parent == rs {
+			server.unsubscribe()
+			s.eventLock.Lock()
+			s.addEvent(Event{Type: EvUnregistered, Server: server})
+			s.eventLock.Unlock()
+			delete(s.servers, server)
+			return
+		}
+	}
+}
+
+// Start starts the scheduler. It should be called after registering all modules
+// and before registering any servers.
+func (s *Scheduler) Start() {
+	go s.syncLoop()
+}
+
+// Stop stops the scheduler.
+func (s *Scheduler) Stop() {
+	s.lock.Lock()
+	for server := range s.servers {
+		server.unsubscribe()
+	}
+	s.servers = nil
+	s.lock.Unlock()
+	stop := make(chan struct{})
+	s.stopCh <- stop
+	<-stop
+}
+
+// syncLoop calls all modules in the order of their registration.
+// A round of processing starts whenever the global trigger is fired. Triggers
+// fired during a processing round ensure that there is going to be a next round.
+func (s *Scheduler) syncLoop() {
+	for {
+		s.lock.Lock()
+		for {
+			s.processModules()
+			if !s.targetChanged() {
+				break
+			}
+		}
+		s.sendRequests()
+		s.lock.Unlock()
+	loop:
+		for {
+			select {
+			case stop := <-s.stopCh:
+				close(stop)
+				return
+			case <-s.triggerCh:
+				break loop
+				//case <-s.testWaitCh:
+			}
+		}
+	}
+}
+
+func (s *Scheduler) targetChanged() (changed bool) {
+	for target, counter := range s.targets {
+		if newCounter := target.ChangeCounter(); newCounter != counter {
+			s.targets[target] = newCounter
+			changed = true
+		}
+	}
+	return
+}
+
+// processModules runs an entire processing round, calling the Process functions
+// of all modules, passing all relevant events.
+func (s *Scheduler) processModules() {
+	s.eventLock.Lock()
+	serverEvents, requestEvents := s.serverEvents, s.requestEvents
+	s.serverEvents, s.requestEvents = nil, nil
+	s.eventLock.Unlock()
+
+	log.Debug("Processing modules", "server events", len(s.serverEvents))
+	for _, module := range s.modules {
+		log.Debug("Processing module", "name", s.names[module], "request events", len(s.requestEvents[module]))
+		module.Process(append(serverEvents, requestEvents[module]...))
+	}
+}
+
+func (s *Scheduler) sendRequests() {
+	servers := make(serverSet)
+	for server := range s.servers {
+		if ok, _ := server.canRequestNow(); ok {
+			servers[server] = struct{}{}
+		}
+	}
+	log.Debug("Generating request candidates", "servers", len(servers))
+
+	for _, module := range s.modules {
+		if len(servers) == 0 {
+			return
+		}
```

Comment:
It makes sense to check here because `tryRequest` can remove servers from the map if they cannot accept any more requests at the moment. If there are no more servers to request from then there is no need to iterate the modules.